### ISSUE TITLE
Changing center of mass of a body, new Inertia, Parallel Axis Theorem

### ISSUE THREAD

**Issue by gyakoo** [erincatto/box2d#521]
Maybe I'm wrong, if that's the case please feel free to close it.

I'm using the following **game code** to update the center of mass of a body:

```
if ( bd.type == b2_dynamicBody  )
{
    b2MassData massData;
    if ( bodyDef.density <= SYS_EPSILON )
    {
        // no density valid, use manual masses
        assert( bodyDef.mass > 0.0f );
        assert( bodyDef.inertia > 0.0f );
        massData.mass = bodyDef.mass;
        massData.I = bodyDef.inertia;
    }
    else
    {
        // density was valid, so there's mass data, just get it
        body->GetMassData( &massData );
    }

    // and set the center of mass
    GAME_TO_B2( bodyDef.centerLocal, massData.center );
    body->SetMassData( &massData );
}
```

The problem is that it **asserts in b2Body.cpp**, in the block:

```
if (massData->I > 0.0f && (m_flags & b2Body::e_fixedRotationFlag) == 0)
{
	m_I = massData->I - m_mass * b2Dot(massData->center, massData->center);
	b2Assert(m_I > 0.0f);
	m_invI = 1.0f / m_I;
}
```

If I understand correctly, it tries to compute the new Inertia through the new center of mass, and it looks the Parallel Axis Theorem. My question is _shouldn't the subtract be an addition instead_ so it becomes:
```
	m_I = massData->I + m_mass * b2Dot(massData->center, massData->center);
```
By changing it, we can remove the assert and the behaviour that I get is correct.

Any thoughts? 

Thanks!

**Comment by mcourteaux** [comment]
I think it shouldn't. Look at the graph drawn on Wikipedia for the PAT. `d` measures the distance between the new axis and the original axis that goes through the center of mass. But the code says: `b2Dot(center, center)`. So this is the distance of the axis (now coming out of your screen, as it is 2D), to the local origin of the body definition. So the `massData->center` is the real center of mass.

Basically in the formula of Wikipedia: `I = I_cm + m*d^2`, you are confusing `I` with `I_cm`.

**Comment by erincatto** [comment]
Yeah, the inertia in b2MassData is about the local origin. I recommend to first take the result from GetMassData and convert the inertia to the old center of mass.
Ic = I1 - m * dot(c1, c1)
Then compute the new inertia:
I2 = Ic + m * dot(c2, c2)
Then Box2D will find Ic correctly.

**Comment by luisgrimaldo** [comment]
@erincatto you mean that when I want to only change the center of mass, without changing the mass, I'd have to do something like this:
```
b2MassData data;
_b2Body->GetMassData(&data);
data.center = centerOfMass;
data.I = data.I + data.mass * b2Dot(data.center, data.center);
_b2Body->SetMassData(&data);
```
I tried it and it doesn't seem to work.
I also tried adding all my fixtures' shapes rotational inertias and using that as `lc` but it still doesn't work.

**Comment by HughPH** [comment]
My understanding is that Box2D calculates the centre of mass from the dimensions of the shape - and it is therefore immutable. The density is assumed to be constant throughout the body.

To achieve what you want, you might think about this in physical terms: you could give the body a lower density and attach a small circular body with a high density at the point that you want the centre of mass to be. A bit like placing a penny on the weights in the Big Ben clock, or gluing weights to the inside of a car wheel to account for manufacturing tolerance in tyres.
